text/0000-fragment-specifiers-for-generic-arguments.md

```diff
@@ -0,0 +1,155 @@
+
+# Motivation
+[motivation]: #motivation
+
+I personally encountered this issue when attempting to write a declarative macro
+to implement a trait on a type. I wanted to write something like this minimal
+toy example:
+
+```rust
+macro_rules! implement_debug {
+    { $params:generic $type:ty } => {
+        impl $params Debug for $ty {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+```
+
+However, with the current state of declarative macros, there's no way to parse
+arbitrary generic parameters in the body of the macro, forcing this macro of
+mine to be a procedural macro. However, declarative macros are easier to read
+and write, and this could be a declarative macro if there was only a way to
+parse generic parameters.
+
+Additionally, more complicated macros want to be able to parse each parameters
+and its bounds for use in various places, so I'd like this as well:
+
+```rust
+macro_rules! implement_debug {
+    { < $( $param:generic_param $( : $bound:generic_bound )? ),+ > $type:ty } => {
+        impl < $( $param $( : $bound )? ),+ > Debug for $ty {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+```
+
+Any toy example will be obviously redundant with `:generic`, but more involved
+macros sometimes want it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When explaining fragment specifiers, this can be explained by adding the new
+fragment specifiers and their descriptions:
+
+* `:generic`: A full set of generic parameters and their bounds (e.g. `<'a, T:
+  'a + SomeTrait, const N: usize>`)
+* `:generic_param`: A generic parameter (e.g. `'a`, `T`, or `const N`)
+* `:generic_bound`: Bounds on a generic parameter (e.g. `'lifetime + SomeTrait`
+  on a type or `usize` on a const parameter).
+* `:generic_default`: A default value for a generic type or lifetime
+
+These four parameters are designed to make it easier to write declarative macros
+that take in generic arguments (e.g. to use with a type or function), and then
+use them to be generic on e.g. type definitions or `impl` blocks.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Exact parsing behavior:
+* `:generic` matches the
+  [`GenericParams`](https://doc.rust-lang.org/reference/items/generics.html)
+  grammar item.
+* `:generic_param` matches any of a lifetime, an identifier, or `const` followed
+  by an identifier.
+* `:generic_bound` matches the
+  [`TypeParamBounds`](https://doc.rust-lang.org/reference/trait-bounds.html)
+  (can be the bounds on a type parameter) or
+  [`LifetimeBounds`](https://doc.rust-lang.org/reference/trait-bounds.html) (can
+  be the bounds on a lifetime parameter) grammar items, or a type (can be the
+  bounds on a const parameter).
```

Comment:
Should it be named "generic_bound" or "generic_bound**s**"?
Also that `TypeParamBounds` does not allow empty, while the generic syntax do allow `<T: /*nothing*/>` (except for const generics). If we want it to match multiple bounds, it should also match no bounds.

Reply:
Neat, I never knew that Rust allowed `<T:>` as a generic parameter list. That doesn't seem very useful to me, but I'll change it to allow that since I agree that it should be matched if Rust allows it.